emission/tests/modellingTests/TestSimilarityMetric.py

```diff
@@ -6,26 +6,93 @@ class TestSimilarityMetric(unittest.TestCase):
+        coords1 = metric.extract_features(trips[1])[:2]        
+        similarO = metric.similar(coords0, coords1, similarity_threshold)
+
+        ##Sub-Test 5 :
+        # random, but, only destination points are sampled within a circle and should always be < sim threshold
+        # Since destination of two points lies within threshold limits,they should be similar
+        # when we check just destination for similarity.
+
+        trips = etmm.generate_mock_trips('Caty', 2, [0, 0], [1, 1], '_d',threshold=generate_points_thresh)        
+        coords0 = metric.extract_features(trips[0])[2:]
+        coords1 = metric.extract_features(trips[1])[2:]        
+        similarD = metric.similar(coords0, coords1, similarity_threshold)
+
+        # All the similars must be true
+        self.assertTrue(similarOD1) # RESULT SUB-TEST 1
+        self.assertTrue(similarOD2) # RESULT SUB-TEST 2
+        self.assertTrue(similarOD3) # RESULT SUB-TEST 3
+        self.assertTrue(similarO)  # RESULT SUB-TEST 4
+        self.assertTrue(similarD) # RESULT SUB-TEST 5
     
     def testODsAreNotSimilar(self):
         generate_points_thresh = 0.001  # approx. 111 meters
-        similarity_threshold = 500  # 
-        
-        trips0 = etmm.generate_mock_trips('bob', 1, [0, 0], [1, 1], threshold=generate_points_thresh)
-        trips1 = etmm.generate_mock_trips('alice', 1, [2, 2], [3, 3], threshold=generate_points_thresh)
+        similarity_threshold = 111  # 
+        metric = eamso.OriginDestinationSimilarity()
+
+        ## Sub-Test 1-2: 
+        # Two trips with neither origin nor destination coordinates within threshold
+        # must not be similar in any configuration of similarity testing.
+        trips = etmm.generate_mock_trips('bob', 2, [0, 0], [1, 1], '__', threshold=generate_points_thresh)  
+        coords0 = metric.extract_features(trips[0])
+        coords1 = metric.extract_features(trips[1])
+        similar11 = metric.similar(coords0[:2], coords1[:2], similarity_threshold)
+        similar12 = metric.similar(coords0[2:], coords1[:], similarity_threshold)
+
+        ## Sub-Test 3-4: 
+        # Two trips with  origin coordinates within threshold but we check  
+        # similarity using destination coordinates or origin-and-destination
+        # should not be similar.
+        trips = etmm.generate_mock_trips('Alice', 2, [2, 2], [3, 3], 'o_', threshold=generate_points_thresh)
         metric = eamso.OriginDestinationSimilarity()
-        coords0 = metric.extract_features(trips0[0])
-        coords1 = metric.extract_features(trips1[0])
-        similar = metric.similar(coords0, coords1, similarity_threshold)
-        self.assertFalse(similar)
+        coords0 = metric.extract_features(trips[0])
+        coords1 = metric.extract_features(trips[1])
+        similar21 = metric.similar(coords0[2:], coords1[2:], similarity_threshold)
+        similar22 = metric.similar(coords0, coords1, similarity_threshold)
+
```

Comment:
I am generally confused here. Both of these use the same metric.
Where are you specifying that we should use destination or origin or both?

Reply:
If both the list have 4 values ( 2 for origin and 2 for destinaiton), the function will treat this as a  o-d similarity and check for similarity in both origin and destination.
if the list has 2 , its either O or D. From the POV of the function, it doesn't need to know if the 2 points are for O or from D , it just needs to tell if 2 points are close enough. The distinction between O and D is upstream, where we decide whether to send first 2 coords ( for origin) or last two (for destinaiton).

Reply:
I understand what you are doing with the arrays.
However, you are not passing in anything to `similar` which indicates whether you should be using `o`, `d` or `od`.

I think you say that the determination of which metric to use is implicit in the array size. In general, I think that explicit is better than implicit here, although I am open to discussion around it.

Reply:
Yes. Its implicit. This similar  to what I am doing in Line 215 in ` greedy_similarity_binning.py` above ( around this [comment ]( https://github.com/e-mission/e-mission-server/pull/933/files#r1303879436)) where I decide the start and end point and then call similarity using if-else. 

Even if I do pass any of the three  (`origin`,`destination`,`origin-destination`) to `similar`,  the way `similar` is written in `od_similarity.py`  ( around this [comment]( https://github.com/e-mission/e-mission-server/pull/933/files#r1303869440)), It'll either :

a. not be of any  use there, OR
b. An if-else will have to be written, which is not required since we discussed previously  moved that decision making upstream to `greedy_similarity_binning.py`.

Alternately, I can add a comment here to explain this implicitness.

Reply:
`GreedySimilaritityBinning` now accepts the `clusteringWay` parameter
and then instead of passing it through to `similar`, it passes in the subset of features instead.

it makes sense that we don't want to have the o, d, od distinction in `similarity_metric` since the metric only cares about whether elements are similar or not.

however, it seems like in `od_similarity`, we are implicitly doing an `o`, `d`, `od` check just in a fairly convoluted fashion by expecting a certain type of input. In general, (this is more art than science), explicit is better than implicit, especially when the inputs are standard values like locations. So I think that `od_similarity` should continue to take full locations but should also take in an explicit `clusteringWay` and do the manipulation into the feature selection in there. If it were not called `od_similarity` maybe I would feel differently but it is and so we should handle the `od` distinction there.

Reply:
We had a discussion regarding this here :

https://github.com/e-mission/e-mission-server/pull/933#pullrequestreview-1588325028

where we decided to move the `clustering_way` based decision making upstream .

Reply:
That discussion was focused on making sure that we did not have the `if-else` in the `similar` function instead the similarity metric. It did not discuss the `od_similarity` module and the `similarity` interface. After looking through the code again, and looking at the interface exposed by the unit tests, you should have the `clusteringWay` passed in to `similarity`, do the if/then inside `similarity` and invoke `similarity` directly from the unit tests.

That will dramatically simplify your unit test interface and make everything much cleaner.